flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionDriverTest.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.highavailability;
+
+import org.apache.flink.kubernetes.kubeclient.FlinkKubeClient;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesConfigMap;
+import org.apache.flink.runtime.leaderelection.LeaderInformation;
+
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.apache.flink.kubernetes.utils.Constants.LABEL_CONFIGMAP_TYPE_HIGH_AVAILABILITY;
+import static org.apache.flink.kubernetes.utils.Constants.LABEL_CONFIGMAP_TYPE_KEY;
+import static org.apache.flink.kubernetes.utils.Constants.LEADER_ADDRESS_KEY;
+import static org.apache.flink.kubernetes.utils.Constants.LEADER_SESSION_ID_KEY;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for the {@link KubernetesLeaderElectionDriver}.
+ */
+public class KubernetesLeaderElectionDriverTest extends KubernetesHighAvailabilityTestBase {
+
+	@Test
+	public void testIsLeader() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					// Grant leadership
+					leaderCallbackGrantLeadership();
+					assertThat(electionEventHandler.isLeader(), is(true));
+					assertThat(electionEventHandler.getConfirmedLeaderInformation(), is(LEADER_INFORMATION));
+				});
+		}};
+	}
+
+	@Test
+	public void testNotLeader() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					leaderCallbackGrantLeadership();
+					// Revoke leadership
+					getLeaderCallback().notLeader();
+
+					electionEventHandler.waitForRevokeLeader(TIMEOUT);
+					assertThat(electionEventHandler.isLeader(), is(false));
+					assertThat(electionEventHandler.getConfirmedLeaderInformation(), is(LeaderInformation.empty()));
+					// The ConfigMap should also be cleared
+					assertThat(getLeaderConfigMap().getData().get(LEADER_ADDRESS_KEY), is(nullValue()));
+					assertThat(getLeaderConfigMap().getData().get(LEADER_SESSION_ID_KEY), is(nullValue()));
```

Comment:
How will this work if the `KubernetesLeaderElectionDriver` checks whether one is currently the leader before writing the leader information?

Reply:
Because we do not have new leader elected and the lock identity annotation on the ConfigMap is still the old one. So the `KubernetesLeaderElector#hasLeadership` could pass the check. I think this is the expected behavior. Once the leader ConfigMap is over written by a new leader. It should return false.

Reply:
ah ok. Thanks for the explanation.